load-test-framework/src/main/java/com/google/pubsub/flic/controllers/Controller.java

```diff
@@ -0,0 +1,122 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+package com.google.pubsub.flic.controllers;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.pubsub.flic.common.LatencyDistribution;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.stream.Collectors;
+
+public abstract class Controller {
+  final static Logger log = LoggerFactory.getLogger(Controller.class);
+  final List<Client> clients = new ArrayList<>();
+  final ScheduledExecutorService executor;
+
+  /*
+  Creates the given environments and starts the virtual machines. When this function returns, each client is guaranteed
+  to have been connected and be network reachable, but is not started. If an error occurred attempting to start the
+  environment, the environment will be shut down, and an IOException will be thrown. It is not guaranteed that we have
+  completed shutting down when this function returns, but it is guaranteed that we are in process.
+   */
+  Controller(ScheduledExecutorService executor) {
+    this.executor = executor;
+  }
+
+  protected abstract void shutdown(Throwable t);
+
+  private Result resultsForType(Client.ClientType type, boolean wait) throws Throwable {
```

Comment:
Instead of having a wait parameter and calling now() on the local controller (which may be different from the clock on the remote VM), the VM server should respond with a running_time in the CheckResponse (which is independent of the 'done' bit in the CheckResponse). This way, the throughput calculation will be very simple and getResults can be invoked in the same way, regardless of whether the VM is done generating its load.